documentation/load-bal-service/fast-path-icmp-flow-redirection.md

```diff
@@ -0,0 +1,180 @@
+    IPV6_HEADER Ipv6Header;
+    uint16 SourcePort;
+    uint16 DestinationPort;
+    uint32 SequenceNumber;
+} RedirectOption;
+
+struct 
+{ 
+    uint32 Version; 
+    uint16 AddrFamily; 
+    uint16 EncapType; 
+    uint32 EncapId; 
+    union { 
+        struct { 
+            in_addr DipPAv4; 
+            char VMMac[MAC_ADDR_SIZE]; 
+        } Info4; 
+        struct { 
+            in6_addr DipPAv6; 
+            char VMMac[MAC_ADDR_SIZE]; 
+        } Info6; 
+} RedirectInfo; 
+```
+
+The following shall be used for translations:
+
+| Field                         | Mapping                       |
+| ----------------------------- | ----------------------------- |
+| VM Mac                        | Source ENI                    |
+| Inner Src IP                  | Original Src IP               |
+| Inner Dst IP                  | Original Dst IP               |
+| Target Address                | Original Dst IP               |
+| Redirect Header               | Original IPv6 Header + TCP ports (5 tuple) |
+| Addr Family                   | AF_INET/AF_INET6              |
+| Encap Type                    | NVGRE 1/VXLAN 2               |
+| Encap Id                      | Redirect GRE Key/ VXLAN Id    |
+| Custom Redirect Info          | Redirect DIP and Dst Mac      |
+
+### Packet signatures from captures
+
+- TCP handshake uses VIP (52.184.168.32)
+
+  ![TCP-handshake](images/TCP-handshake.png)
+
+- ICMP Redirect packet
+
+  ![ICMP-redirect.png](images/ICMP-redirect.png)
+
+- After the ICMP redirect, the packets start using DIP/PA (100.110.225.76)
+
+  ![after-ICMP-redirect](images/after-ICMP-redirect.png)
+
+## Detailed design
+
+### SAI API for fast path implementation
+
+All fast path packet handling and flow manipulation are done under the SAI API, hence there is no SAI API added for implementing the feature.
+
+### SAI API for enabling and disabling fast path
+
+In case of live sites, for example, fast path goes wrong and incorrectly updates the flow, we will need a toggle to disable this behavior.
+
+This toggle is added as an attribute to ENI:
+
+```c
+typedef enum _sai_eni_attr_t
+{
+    // ...
+
+    /**
+     * @brief Action set_eni_attrs parameter FAST_PATH_ICMP_FLOW_REDIRECTION
+     *
+     * @type bool
+     * @flags CREATE_AND_SET
+     * @default false
+     */
+    SAI_ENI_ATTR_DISABLE_FAST_PATH_ICMP_FLOW_REDIRECTION,
+
+    // ...
+} sai_eni_attr_t;
+```
+
+### Fast Path counters
+
+For debugging purposes, we will also need to add counters in order to give insights on how fast path works internally.
+
+#### Port level counter
+
+Port level counter will be added as port attribute extensions following the [SAI extension model](https://github.com/opencomputeproject/SAI/blob/master/doc/SAI-Extensions.md#extension-custom-attributes).
+
+| Attribute name | Description |
+| -------------- | ----------- |
+| SAI_PORT_ATTR_FAST_PATH_ICMP_IN_COUNTER_ID | The number of fast path packets/bytes received |
```

Comment:
@marian-pritsak , do you have any concern on collecting this information on port level?

(btw, this will be changed to use sai_port_stat_t instead of counters, due to resource usage concerns.)

Reply:
Need to discuss to determine the best place for this per Marian & Riff

Reply:
discussed with Marian and Prince offline and updated the doc accordingly.